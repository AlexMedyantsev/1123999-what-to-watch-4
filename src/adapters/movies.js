const getLevel = (score) => {
  let level = ``;

  if (score < 3) {
    level = `Bad`;
  } else if (score >= 3 && score < 5) {
    level = `Normal`;
  } else if (score >= 5 && score < 8) {
    level = `Good`;
  } else if (score >= 8 && score < 10) {
    level = `Very Good`;
  } else if (score === 10) {
    level = `Awesome`;
  }

  return level;
};

export const adaptMovies = (moviesFromServer) => {
  return moviesFromServer.map((movie) => {
    return ({
      image: movie.preview_image,
      posterSrc: movie.poster_image,
      bgSrc: movie.background_image,
      title: movie.name,
      genre: movie.genre,
      year: movie.released,
      score: movie.rating,
      level: getLevel(movie.rating),
      movieLink: movie.video_link,
      previewLink: movie.preview_video_link,
      scoresCount: movie.scores_count,
      description: movie.description,
      director: movie.director,
      starring: movie.starring,
      runTime: movie.run_time,
      link: `movie-page.html`,
      id: movie.id,
      backgroundColor: movie.background_color,
      isFavorite: movie.is_favorite
    });
  });
}

export const adaptMovie = (movie) => {
  return ({
    image: movie.preview_image,
    posterSrc: movie.poster_image,
    bgSrc: movie.background_image,
    title: movie.name,
    genre: movie.genre,
    year: movie.released,
    score: movie.rating,
    level: getLevel(movie.rating),
    movieLink: movie.video_link,
    previewLink: movie.preview_video_link,
    scoresCount: movie.scores_count,
    description: movie.description,
    director: movie.director,
    starring: movie.starring,
    runTime: movie.run_time,
    link: `movie-page.html`,
    id: movie.id,
    backgroundColor: movie.background_color,
    isFavorite: movie.is_favorite
  });
};
