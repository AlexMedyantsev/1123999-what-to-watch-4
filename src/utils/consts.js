export const VideoPreview = {
  width: 280,
  height: 175,
  IS_MUTED: true,
  INTERVAL_IN_SEC: 1000,
};

export const PREVIEW_PLAY_DELAY = 2000;

export const MovieDetailsTabs = {
  OVERVIEW: `Overview`,
  DETAILS: `Details`,
  REVIEWS: `Reviews`,
};

export const GENRES = {
  ALL: `All genres`,
  ADVENTURE: `Adventure`,
  COMEDIES: `Comedy`,
  CRIME: `Crime`,
  DOCUMENTARY: `Documentary`,
  DRAMAS: `Drama`,
  FANTASY: `Fantasy`,
  HORROR: `Horror`,
  KIDS_AND_FAMILY: `Kid's and Family`,
  ROMANCE: `Romance`,
  SCI_FI: `Sci - Fi`,
  STORY: `Story`,
  THRILLERS: `Thriller`,
};

export const MAX_NUMBER_GENRES = 9;

export const FILTER_ALL_GENRES = `All Genres`;

export const SHOWING_MOVIES_COUNT_ON_START = 8;

export const SIMILAR_MOVIES_COUNT = 4;

export const MOVIES_ADDED_ON_BUTTON_CLICK = 8;

export const ESCAPE_KEY = 27;
