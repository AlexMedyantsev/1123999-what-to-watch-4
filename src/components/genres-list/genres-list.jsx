import React, {PureComponent} from "react";
import {ActionCreator} from "../../reducer.js";
import {connect} from "react-redux";
import PropTypes from "prop-types";
import {MAX_NUMBER_GENRES} from "../../utils/consts.js";

class GenresList extends PureComponent {
  constructor(props) {
    super(props);

    this._getGenresList = this._getGenresList.bind(this);
  }

  _getGenresList() {
    const genres = new Set();

    genres.add(`All genres`);

    this.props.movies.forEach((movie) => {
      movie.genres.forEach((genre) => genres.has(genre) ? `` : genres.add(genre));
    });

    const genresList = Array.from(genres).slice(0, MAX_NUMBER_GENRES);
    return genresList;
  }

  render() {
    const {genre, onGenreClick} = this.props;
    const genresList = this._getGenresList();

    return (
      <ul className="catalog__genres-list">
        {genresList.map((genreItem) => {
          const activeClass = genre === genreItem ? `catalog__genres-item--active` : ``;

          return <li
            key={genreItem}
            className={`catalog__genres-item ${activeClass}`}
            onClick={(evt)=>{
              evt.preventDefault();
              onGenreClick(genreItem);
            }}
          >
            <a href="#" className="catalog__genres-link">{genreItem}</a>
          </li>;
        })}
      </ul>
    );
  }
}

GenresList.propTypes = {
  genre: PropTypes.string.isRequired,
  onGenreClick: PropTypes.func,
  movies: PropTypes.arrayOf(
      PropTypes.shape({
        id: PropTypes.number,
        title: PropTypes.string,
        image: PropTypes.string.isRequired,
      }).isRequired
  ).isRequired,
};

const mapStateToProps = (state) => ({
  genre: state.genre,
});

const mapDispatchToProps = (dispatch) => ({
  onGenreClick(genre) {
    dispatch(ActionCreator.setActiveGenre(genre));
  },
});

export {GenresList};
export default connect(mapStateToProps, mapDispatchToProps)(GenresList);
