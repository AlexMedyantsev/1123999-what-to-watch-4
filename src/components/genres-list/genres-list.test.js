import React from "react";
import renderer from "react-test-renderer";

import {GenresList} from "./genres-list.jsx";

import {GENRES} from "../../utils/consts.js";
import movies from "../../mocks/movies.js";


describe(`Main Snapshot`, () => {
  it(`Should GenresList render correctly`, () => {
    const tree = renderer
        .create(
            <GenresList
              genre={GENRES.ALL}
              movies={movies}
              onClick={()=>{}}
            />)
        .toJSON();

    expect(tree).toMatchSnapshot();
  });
});
