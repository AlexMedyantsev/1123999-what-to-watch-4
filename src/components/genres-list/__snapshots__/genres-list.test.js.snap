// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Main Snapshot Should GenresList render correctly 1`] = `
<ul
  className="catalog__genres-list"
>
  <li
    className="catalog__genres-item catalog__genres-item--active"
    onClick={[Function]}
  >
    <a
      className="catalog__genres-link"
      href="#"
    >
      All genres
    </a>
  </li>
  <li
    className="catalog__genres-item "
    onClick={[Function]}
  >
    <a
      className="catalog__genres-link"
      href="#"
    >
      Fantasy
    </a>
  </li>
  <li
    className="catalog__genres-item "
    onClick={[Function]}
  >
    <a
      className="catalog__genres-link"
      href="#"
    >
      Kid's and Family
    </a>
  </li>
  <li
    className="catalog__genres-item "
    onClick={[Function]}
  >
    <a
      className="catalog__genres-link"
      href="#"
    >
      Adventure
    </a>
  </li>
  <li
    className="catalog__genres-item "
    onClick={[Function]}
  >
    <a
      className="catalog__genres-link"
      href="#"
    >
      Story
    </a>
  </li>
  <li
    className="catalog__genres-item "
    onClick={[Function]}
  >
    <a
      className="catalog__genres-link"
      href="#"
    >
      Drama
    </a>
  </li>
  <li
    className="catalog__genres-item "
    onClick={[Function]}
  >
    <a
      className="catalog__genres-link"
      href="#"
    >
      Biography
    </a>
  </li>
  <li
    className="catalog__genres-item "
    onClick={[Function]}
  >
    <a
      className="catalog__genres-link"
      href="#"
    >
      Music
    </a>
  </li>
  <li
    className="catalog__genres-item "
    onClick={[Function]}
  >
    <a
      className="catalog__genres-link"
      href="#"
    >
      Thriller
    </a>
  </li>
</ul>
`;
